File: 2018/day14.js
/*

--- Day 14: Chocolate Charts ---

You finally have a chance to look at all of the produce moving around. Chocolate, cinnamon, mint, chili peppers, nutmeg, vanilla... the Elves must be growing these plants to make hot chocolate! As you realize this, you hear a conversation in the distance. When you go to investigate, you discover two Elves in what appears to be a makeshift underground kitchen/laboratory.

The Elves are trying to come up with the ultimate hot chocolate recipe; they're even maintaining a scoreboard which tracks the quality score (0-9) of each recipe.

Only two recipes are on the board: the first recipe got a score of 3, the second, 7. Each of the two Elves has a current recipe: the first Elf starts with the first recipe, and the second Elf starts with the second recipe.

To create new recipes, the two Elves combine their current recipes. This creates new recipes from the digits of the sum of the current recipes' scores. With the current recipes' scores of 3 and 7, their sum is 10, and so two new recipes would be created: the first with score 1 and the second with score 0. If the current recipes' scores were 2 and 3, the sum, 5, would only create one recipe (with a score of 5) with its single digit.

The new recipes are added to the end of the scoreboard in the order they are created. So, after the first round, the scoreboard is 3, 7, 1, 0.

After all new recipes are added to the scoreboard, each Elf picks a new current recipe. To do this, the Elf steps forward through the scoreboard a number of recipes equal to 1 plus the score of their current recipe. So, after the first round, the first Elf moves forward 1 + 3 = 4 times, while the second Elf moves forward 1 + 7 = 8 times. If they run out of recipes, they loop back around to the beginning. After the first round, both Elves happen to loop around until they land on the same recipe that they had in the beginning; in general, they will move to different recipes.

Drawing the first Elf as parentheses and the second Elf as square brackets, they continue this process:

(3)[7]
(3)[7] 1  0 
 3  7  1 [0](1) 0 
 3  7  1  0 [1] 0 (1)
(3) 7  1  0  1  0 [1] 2 
 3  7  1  0 (1) 0  1  2 [4]
 3  7  1 [0] 1  0 (1) 2  4  5 
 3  7  1  0 [1] 0  1  2 (4) 5  1 
 3 (7) 1  0  1  0 [1] 2  4  5  1  5 
 3  7  1  0  1  0  1  2 [4](5) 1  5  8 
 3 (7) 1  0  1  0  1  2  4  5  1  5  8 [9]
 3  7  1  0  1  0  1 [2] 4 (5) 1  5  8  9  1  6 
 3  7  1  0  1  0  1  2  4  5 [1] 5  8  9  1 (6) 7 
 3  7  1  0 (1) 0  1  2  4  5  1  5 [8] 9  1  6  7  7 
 3  7 [1] 0  1  0 (1) 2  4  5  1  5  8  9  1  6  7  7  9 
 3  7  1  0 [1] 0  1  2 (4) 5  1  5  8  9  1  6  7  7  9  2 

The Elves think their skill will improve after making a few recipes (your puzzle input). However, that could take ages; you can speed this up considerably by identifying the scores of the ten recipes after that. For example:

If the Elves think their skill will improve after making 9 recipes, the scores of the ten recipes after the first nine on the scoreboard would be 5158916779 (highlighted in the last line of the diagram).
After 5 recipes, the scores of the next ten would be 0124515891.
After 18 recipes, the scores of the next ten would be 9251071085.
After 2018 recipes, the scores of the next ten would be 5941429882.

What are the scores of the ten recipes immediately after the number of recipes in your puzzle input?


--- Part Two ---

As it turns out, you got the Elves' plan backwards. They actually want to know how many recipes appear on the scoreboard to the left of the first recipes whose scores are the digits from your puzzle input.

51589 first appears after 9 recipes.
01245 first appears after 5 recipes.
92510 first appears after 18 recipes.
59414 first appears after 2018 recipes.

How many recipes appear on the scoreboard to the left of the score sequence in your puzzle input?

*/

// OPTIONAL VARIABLES
const DISPLAY_EXTRA_INFO = true;

function analyzeGrowingSequence (part, inputStr, DEBUG = false) {
  
  // CONSTANTS
  const NUM_ROUNDS = +inputStr;
  const GOAL_SEQUENCE = inputStr;
  
  // DATA STRUCTURES
  const RECIPES = [ 3, 7 ];

  // INIT
  let [ elf1Idx, elf2Idx ] = [ 0, 1 ];

  // ANALYZE
  const TIME_AT_START = Date.now();
  if (part === 2) console.log('RUNNING PART 2 ANALYSIS (PLEASE WAIT)...');

  while ( part === 1  ? RECIPES.length < NUM_ROUNDS + 10                                      // PART 1: STOP WHEN YOU HAVE ENOUGH RECIPES EQUAL TO
                                                                                              // NUMBER OF ROUNDS + 10 (YOU LOOK AT THE NEXT 10 NUMBERS)

                      : !RECIPES.slice(-GOAL_SEQUENCE.length - 1)                             // PART 2: STOP WHEN SOMEWHERE IN YOUR RECIPES YOU FIND
                                .join('')                                                     // THE GOAL SEQUENCE
                                .includes(GOAL_SEQUENCE)                              
  ) {

    // CALCULATE SCORE FOR NEW RECIPE(S)
    const elf1Score = RECIPES[elf1Idx];
    const elf2Score = RECIPES[elf2Idx];
    const sum = elf1Score + elf2Score;
    const tens = sum >= 10 ? 1 : 0;
    const ones = sum % 10;

    // ADD NEW RECIPE(S) TO THE LIST
    if (tens) RECIPES.push(tens);
    RECIPES.push(ones);

    // MOVE ELF INDICES
    elf1Idx = (elf1Idx + elf1Score + 1) % RECIPES.length;
    elf2Idx = (elf2Idx + elf2Score + 1) % RECIPES.length;

    if (DISPLAY_EXTRA_INFO && DEBUG) {
      console.log(`elf1Idx: ${elf1Idx}, elf2Idx: ${elf2Idx} |`, RECIPES.join(' '));
    }

  }

  if (part === 2) console.log(`(RUN TOOK ${(Date.now() - TIME_AT_START)/1000} SECS)`);

  return part === 1 ? RECIPES.slice(NUM_ROUNDS, NUM_ROUNDS + 10).join('')                     // PART 1: SLICE THE 10 NUMBERS AFTER NUMBER OF ROUNDS
                    : RECIPES.join('').indexOf(GOAL_SEQUENCE);                                // PART 2: FIND NUMBER OF POSITIONS LEFT OF THE GOAL SEQUENCE
}

// TEST CASES

const test = require('./_test');
const testNum = [1];
let input, expected;
const func = analyzeGrowingSequence;
const sortedFunc = (...args) => func(...args).sort();                   // used when the order of the output does not matter
const modFunc = (...args) => func(...args) % 1000000007;                // used when the output is very large
const skippedTests = new Set([  ]);
const lowestTest = 0;
const highestTest = 0;

const fs = require('fs');
const path = require('path');
const DAY_NUM = __filename.split('.js')[0].split('day')[1];
const INPUT_PATH = path.join(__dirname, `day${DAY_NUM}-input.txt`);
const actualInput = fs.readFileSync(INPUT_PATH, 'utf8');
const parseSampleInput = s => s.split('').map(c => c === '\n' ? '\r\n' : c).join('');

const sampleInput = parseSampleInput(
`9`
);

const sampleInput2 = parseSampleInput(
`5`
);

const sampleInput3 = parseSampleInput(
`18`
);

const sampleInput4 = parseSampleInput(
`2018`
);

const sampleInput5 = parseSampleInput(
`51589`
);

const sampleInput6 = parseSampleInput(
`01245`
);

const sampleInput7 = parseSampleInput(
`92510`
);

const sampleInput8 = parseSampleInput(
`59414`
);

// Test case 1
input = {
  part: 1,
  inputStr: sampleInput,
  DEBUG: true,
};
expected = '5158916779';
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 2
input = {
  part: 1,
  inputStr: sampleInput2,
  DEBUG: true,
};
expected = '0124515891';
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 3
input = {
  part: 1,
  inputStr: sampleInput3,
  DEBUG: true,
};
expected = '9251071085';
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 4
input = {
  part: 1,
  inputStr: sampleInput4,
  // DEBUG: true,
  DEBUG: false,
};
expected = '5941429882';
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 5
input = {
  part: 1,
  inputStr: actualInput,
};
expected = '3147574107';
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 6
input = {
  part: 2,
  inputStr: sampleInput5,
  DEBUG: true,
};
expected = 9;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 7
input = {
  part: 2,
  inputStr: sampleInput6,
  DEBUG: true,
};
expected = 5;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 8
input = {
  part: 2,
  inputStr: sampleInput7,
  DEBUG: true,
};
expected = 18;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 9
input = {
  part: 2,
  inputStr: sampleInput8,
  // DEBUG: true,
  DEBUG: false,
};
expected = 2018;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 10
input = {
  part: 2,
  inputStr: actualInput,
};
expected = 20280190;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

File: 2023/day14.py
"""

--- Day 14: Parabolic Reflector Dish ---

You reach the place where all of the mirrors were pointing: a massive parabolic reflector dish attached to the side of another large mountain.

The dish is made up of many small mirrors, but while the mirrors themselves are roughly in the shape of a parabolic reflector dish, each individual mirror seems to be pointing in slightly the wrong direction. If the dish is meant to focus light, all it's doing right now is sending it in a vague direction.

This system must be what provides the energy for the lava! If you focus the reflector dish, maybe you can go where it's pointing and use the light to fix the lava production.

Upon closer inspection, the individual mirrors each appear to be connected via an elaborate system of ropes and pulleys to a large metal platform below the dish. The platform is covered in large rocks of various shapes. Depending on their position, the weight of the rocks deforms the platform, and the shape of the platform controls which ropes move and ultimately the focus of the dish.

In short: if you move the rocks, you can focus the dish. The platform even has a control panel on the side that lets you tilt it in one of four directions! The rounded rocks (O) will roll when the platform is tilted, while the cube-shaped rocks (#) will stay in place. You note the positions of all of the empty spaces (.) and rocks (your puzzle input). For example:

O....#....
O.OO#....#
.....##...
OO.#O....O
.O.....O#.
O.#..O.#.#
..O..#O..O
.......O..
#....###..
#OO..#....

Start by tilting the lever so all of the rocks will slide north as far as they will go:

OOOO.#.O..
OO..#....#
OO..O##..O
O..#.OO...
........#.
..#....#.#
..O..#.O.O
..O.......
#....###..
#....#....

You notice that the support beams along the north side of the platform are damaged; to ensure the platform doesn't collapse, you should calculate the total load on the north support beams.

The amount of load caused by a single rounded rock (O) is equal to the number of rows from the rock to the south edge of the platform, including the row the rock is on. (Cube-shaped rocks (#) don't contribute to load.) So, the amount of load caused by each rock in each row is as follows:

OOOO.#.O.. 10
OO..#....#  9
OO..O##..O  8
O..#.OO...  7
........#.  6
..#....#.#  5
..O..#.O.O  4
..O.......  3
#....###..  2
#....#....  1

The total load is the sum of the load caused by all of the rounded rocks. In this example, the total load is 136.

Tilt the platform so that the rounded rocks all roll north. Afterward, what is the total load on the north support beams?


--- Part Two ---

The parabolic reflector dish deforms, but not in a way that focuses the beam. To do that, you'll need to move the rocks to the edges of the platform. Fortunately, a button on the side of the control panel labeled "spin cycle" attempts to do just that!

Each cycle tilts the platform four times so that the rounded rocks roll north, then west, then south, then east. After each tilt, the rounded rocks roll as far as they can before the platform tilts in the next direction. After one cycle, the platform will have finished rolling the rounded rocks in those four directions in that order.

Here's what happens in the example above after each of the first few cycles:

After 1 cycle:
.....#....
....#...O#
...OO##...
.OO#......
.....OOO#.
.O#...O#.#
....O#....
......OOOO
#...O###..
#..OO#....

After 2 cycles:
.....#....
....#...O#
.....##...
..O#......
.....OOO#.
.O#...O#.#
....O#...O
.......OOO
#..OO###..
#.OOO#...O

After 3 cycles:
.....#....
....#...O#
.....##...
..O#......
.....OOO#.
.O#...O#.#
....O#...O
.......OOO
#...O###.O
#.OOO#...O

This process should work if you leave it running long enough, but you're still worried about the north support beams. To make sure they'll survive for a while, you need to calculate the total load on the north support beams after 1000000000 cycles.

In the above example, after 1000000000 cycles, the total load on the north support beams is 64.

Run the spin cycle for 1000000000 cycles. Afterward, what is the total load on the north support beams?


"""

# LIBRARIES
import time
import pathlib
import os
from math import *
from functools import *
from sys import *
from collections import *

# MODULES
from _test import test

# OPTIONAL VARIABLES
DISPLAY_EXTRA_INFO = True
# DISPLAY_EXTRA_INFO = False

def roll_rocks_in_cardinal_direction(part, input_str, DEBUG = False):

  # PARSE INPUT DATA

  input_arr = input_str.split('\n')
  MAP = [ [ c for c in line ] for line in input_arr ]


  # DATA STRUCTURES

  MEMO = {}                                                               # keys are serialized positions, vals are the index of first appearance
  LOAD_VALUE = {}                                                                # keys are indices, vals are load values


  # CONSTANTS

  H, W = len(MAP), len(MAP[0])
  ROCK, CUBE, EMPTY = 'O', '#', '.'
  n, e, s, w = 'N', 'E', 'S', 'W'                                         # CAREFUL! W (capital) is already being used

  PARAMS = {
    n: {
      'r_range': range(1, H),
      'c_range': range(W),
      'deltas': (-1, 0),
    },
    w: {
      'r_range': range(H),
      'c_range': range(1, W),
      'deltas': (0, -1),
    },
    s: {
      'r_range': range(H - 1, -1, -1),
      'c_range': range(W),
      'deltas': (+1, 0),
    },
    e: {
      'r_range': range(H),
      'c_range': range(W - 1, -1, -1),
      'deltas': (0, +1),
    },
  }


  # HELPER FUNCTIONS

  def roll(dir):

    assert dir in (n, w, e, s)
    params = PARAMS[dir]

    for r in params['r_range']:                                           # iterate through the relevant rows...
      for c in params['c_range']:                                         # ...and cols...
        if MAP[r][c] == ROCK:
          dy, dx = params['deltas']                                       # (of course, one of these will be 0)
          new_r, new_c = r, c                                             # new_r, new_c represents potential destination to which a rock will roll
          while 0 <= (new_r + dy) < H \
                and 0 <= (new_c + dx) < H \
                and MAP[new_r + dy][new_c + dx] == EMPTY:                 # check if destination would be in bounds and empty
            new_r += dy
            new_c += dx
          MAP[r][c] = EMPTY                                               # once the while loop is done, empty the original rock location...
          MAP[new_r][new_c] = ROCK                                        # ...and place the rock at the new location (which may be the same location)

  def serialize():
    return ''.join([ ''.join(row) for row in MAP ])                       # convert entire board position into one big string

  def roll_cycle(i):

    if DEBUG and 0 < i <= 3 and DISPLAY_EXTRA_INFO:                       # for part 2 sample test, show the results after the first 3 cycles
      print(f"MAP AFTER {i} CYCLE(S):")
      for row in MAP:
        print(''.join(row))
      print('')

    # Check for repeated position
    serial = serialize()
    if serial in MEMO:
      return (True, serial)                                               # output[0] is whether a cycle was found; output[1] is the serialized board position

    # Cache current position
    MEMO[serial] = i
    LOAD_VALUE[i] = get_load()

    # Perform spin cycle, then return False to finish the iteration
    roll(n)
    roll(w)
    roll(s)
    roll(e)
    return (False, serial)

  def get_load():
    output = 0
    for r in range(H):
      output += sum(map(lambda x: (H - r) if x == ROCK else 0, MAP[r]))
    return output


  # ANALYZE

  TIME_AT_START = time.time()

  if part == 1:                                                           # PART 1: DO A SINGLE ROLL NORTH

    roll(n)
    return get_load()

  else:                                                                   # PART 2: DO 1,000,000,000 FULL CYCLES (n, w, s, e)

    if not DEBUG: print('RUNNING PART 2 ANALYSIS (PLEASE WAIT)...')

    NUM_CYCLES = 1000000000
    output = None
    for i in range(NUM_CYCLES):
      found_cycle, serial = roll_cycle(i)
      if found_cycle:
        period = i - MEMO[serial]
        initial_lead = MEMO[serial]
        MOD = (NUM_CYCLES - initial_lead) % period
        output = LOAD_VALUE[initial_lead + MOD]
        if DISPLAY_EXTRA_INFO:
          print(f"Current position after {i} rolls (i.e. before the roll on i == {i}) is a repeat of same on i == {MEMO[serial]}.")
          print(f"Therefore, period is {i} - {MEMO[serial]} = {period}.")
          print(f"The first {initial_lead} cycles, before the first occurrence of the repeated position, is not part of the loop.")
          print(f"After the initial lead of {initial_lead}, if we skip all segments of {period}, a remainder of {MOD} will be left over before we reach {NUM_CYCLES}.")
          print(f"Thus, the result after {NUM_CYCLES} cycles is equal to the result after only {initial_lead} + {MOD} = {initial_lead + MOD} cycles.")
          print(f"Therefore, the answer is equal to the result at i == {initial_lead + MOD}.")
          print(f"LOAD_VALUE[{initial_lead + MOD}] == {output}.")
        break

    assert output

    if not DEBUG: print(f"(RUN TOOK {(time.time() - TIME_AT_START)} SECS)")
    return output


# TEST CASES

test_num = [1]
test_input = None
test_expected = None
func = roll_rocks_in_cardinal_direction
skipped_tests = set([ 2, 3, 4 ])
skipped_tests = set([ 3, 4 ])
skipped_tests = set([ 4 ])
skipped_tests = set([  ])
lowest_test = 0
highest_test = 0

FILENAME = pathlib.Path(__file__).name
FILENAME_SPLIT = FILENAME.split('.py')[0].split('day')
DAY_NUM = FILENAME_SPLIT[1] if len(FILENAME_SPLIT) > 1 else None
CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
INPUT_PATH = CURRENT_DIR + '\\' + (f"day{DAY_NUM}-input.txt" if DAY_NUM else 'template-input.txt')
actual_input = open(INPUT_PATH, 'r').read()

sample_input = """O....#....
O.OO#....#
.....##...
OO.#O....O
.O.....O#.
O.#..O.#.#
..O..#O..O
.......O..
#....###..
#OO..#...."""

# Test case 1
test_input = {
  'part': 1,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 136
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 2
test_input = {
  'part': 1,
  'input_str': actual_input,
}
test_expected = 112048
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 3
test_input = {
  'part': 2,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 64
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 4
test_input = {
  'part': 2,
  'input_str': actual_input,
}
test_expected = 105606
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)